src/ServiceControl/Monitoring/MonitoringDataPersister.cs

```diff
@@ -0,0 +1,101 @@
+    using ServiceControl.EndpointControl.Contracts;
+    using ServiceControl.Infrastructure;
+    using ServiceControl.Infrastructure.DomainEvents;
+
+    public class MonitoringDataPersister : IDomainHandler<HeartbeatingEndpointDetected>,
+        IDomainHandler<MonitoringEnabledForEndpoint>,
+        IDomainHandler<MonitoringDisabledForEndpoint>
+    {
+
+        public void Handle(HeartbeatingEndpointDetected domainEvent)
+        {
+            var id = DeterministicGuid.MakeId(domainEvent.Endpoint.Name, domainEvent.Endpoint.HostId.ToString());
+
+            using (var session = store.OpenSession())
+            {
+                var knownEndpoint = new KnownEndpoint
+                {
+                    Id = id,
+                    EndpointDetails = domainEvent.Endpoint,
+                    HostDisplayName = domainEvent.Endpoint.Host,
+                    Monitored = monitor.IsMonitored(id)
+                };
+
+                session.Store(knownEndpoint);
+
+                session.SaveChanges();
+            }
+        }
+
+        public void Handle(MonitoringEnabledForEndpoint domainEvent)
+        {
+            var id = DeterministicGuid.MakeId(domainEvent.Endpoint.Name, domainEvent.Endpoint.HostId.ToString());
+            using (var session = store.OpenSession())
+            {
+                var knownEndpoint = session.Load<KnownEndpoint>(id);
+                if (knownEndpoint == null)
+                {
+                    knownEndpoint = new KnownEndpoint
+                    {
+                        Id = id,
+                        EndpointDetails = domainEvent.Endpoint,
+                        HostDisplayName = domainEvent.Endpoint.Host,
+                    };
+                    session.Store(knownEndpoint);
+                }
+
+                knownEndpoint.Monitored = true;
+                session.SaveChanges();
+            }
+        }
+
+        public void Handle(MonitoringDisabledForEndpoint domainEvent)
+        {
+            var id = DeterministicGuid.MakeId(domainEvent.Endpoint.Name, domainEvent.Endpoint.HostId.ToString());
+            using (var session = store.OpenSession())
+            {
+                var knownEndpoint = session.Load<KnownEndpoint>(id);
+                if (knownEndpoint == null)
+                {
+                    knownEndpoint = new KnownEndpoint
+                    {
+                        Id = id,
+                        EndpointDetails = domainEvent.Endpoint,
+                        HostDisplayName = domainEvent.Endpoint.Host,
+                    };
+                    session.Store(knownEndpoint);
+                }
+
+                knownEndpoint.Monitored = false;
+                session.SaveChanges();
+            }
+        }
+
+        public void WarmupMonitoringFromPersistence()
+        {
+            using (var session = store.OpenSession())
+            {
+                var endpoints = session.Query<KnownEndpoint>().ToArray();
```

Comment:
This will break if there are more endpoints than the default page size in raven.  Will have to remove the ToArray and switch to an advanced query and stream the results.

```csharp
using (var endpointsEnumerator = session.Advanced.Stream(session.Query<KnownEndpoints>()))
{
    while (endpointsEnumerator.MoveNext())
    {
        var endpoint = endpointsEnumerator.Current.Document;
        monitor.Register(new EndpointInstanceId(endpoint.EndpointDetails.Name, endpoint.EndpointDetails.Host, endpoint.EndpointDetails.HostId), endpoint.Monitored);
    }
}
```